src/GitHub.App/Services/GitHubContextService.cs

```diff
@@ -0,0 +1,474 @@
+                            return (resolveRef, null, commit.Sha);
+                        }
+                    }
+                }
+
+                return (null, null, null);
+            }
+
+            IEnumerable<(string commitish, string path)> ToObjectish(string treeishPath)
+            {
+                var index = 0;
+                while ((index = treeishPath.IndexOf('/', index + 1)) != -1)
+                {
+                    var commitish = treeishPath.Substring(0, index);
+                    var path = treeishPath.Substring(index + 1);
+                    yield return (commitish, path);
+                }
+            }
+        }
+
+        public bool HasChangesInWorkingDirectory(string repositoryDir, string commitish, string path)
+        {
+            using (var repo = gitService.GetRepository(repositoryDir))
+            {
+                var commit = repo.Lookup<Commit>(commitish);
+                var paths = new[] { path };
+
+                return repo.Diff.Compare<Patch>(commit.Tree, DiffTargets.WorkingDirectory, paths).Count() > 0;
+            }
+        }
+
+        public async Task<bool> TryAnnotateFile(string repositoryDir, string currentBranch, GitHubContext context)
+        {
+            var (commitish, path, commitSha) = ResolveBlob(repositoryDir, context);
+            if (path == null)
+            {
+                return false;
+            }
+
+            if (!AnnotateFile(repositoryDir, currentBranch, path, commitSha))
+            {
+                return false;
+            }
+
+            if (context.Line != null)
+            {
+                await Task.Delay(1000);
+                var activeView = FindActiveView();
+                SetSelection(activeView, context);
+            }
+
+            return true;
+        }
+
+        IVsTextView FindActiveView()
+        {
+            if (!ErrorHandler.Succeeded(textManager.Value.GetActiveView2(1, null, (uint)_VIEWFRAMETYPE.vftToolWindow, out IVsTextView textView)))
+            {
+                return null;
+            }
+
+            return textView;
+        }
+
+        bool AnnotateFile(string repositoryPath, string branchName, string relativePath, string versionSha)
+        {
+            var gitExt2Type = Type.GetType("Microsoft.VisualStudio.TeamFoundation.Git.Extensibility.IGitExt2, Microsoft.TeamFoundation.Git.Provider", false);
+            if (gitExt2Type == null)
+            {
+                return false;
+            }
+
+            InvokeService(gitExt2Type, "AnnotateFile", repositoryPath, branchName, relativePath, versionSha);
+            return true;
+        }
+
+        void InvokeService<T1, T2, T3, T4>(Type serviceType, string method, T1 arg1, T2 arg2, T3 arg3, T4 arg4)
+        {
+            var service = serviceProvider.GetService(serviceType);
+            var action = (Action<T1, T2, T3, T4>)Delegate.CreateDelegate(typeof(Action<T1, T2, T3, T4>), service, method);
```

Comment:
Ohh, magic! Could you explain what's happening here? :)

Reply:
The `IGitExt2` interface was introduced in an update of Visual Studio 2017, so we can't be sure it exists. I'm grabbing the `IGitExt2` type by name and if it exists constructing and invoking a delegate on the `AnnotateFile` method.

My helper method was retrieving the service and calling the method. I've simplified it to only create and invoke the delegate. Using reflection was just too ugly. ;-)

The method is now:
```cs
            void Invoke<T1, T2, T3, T4>(object target, string method, T1 arg1, T2 arg2, T3 arg3, T4 arg4)
            {
                var action = (Action<T1, T2, T3, T4>)Delegate.CreateDelegate(typeof(Action<T1, T2, T3, T4>), target, method);
                action.Invoke(arg1, arg2, arg3, arg4);
            }
```